ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -0,0 +1,414 @@
+      = go lastValid (epoch + 1) prevHash
+      | otherwise
+      = assert (epoch == lastEpoch) $ do
+        cleanup lastValid epoch
+        return lastValid
+
+    -- | Remove left over files from epochs newer than the last epoch
+    -- containing a valid file. Also unfinalise it if necessary.
+    cleanup
+      :: (EpochNo, ImmTipWithHash hash)  -- ^ The last valid epoch and tip
+      -> EpochNo  -- ^ The last validated epoch, could have been invalid or
+                  -- empty
+      -> m ()
+    cleanup (lastValidEpoch, tip) lastValidatedEpoch = case tip of
+      TipGen ->
+        removeFilesStartingFrom hasFS 0
+      Tip _  -> do
+        removeFilesStartingFrom hasFS (lastValidEpoch + 1)
+        when (lastValidEpoch < lastValidatedEpoch) $
+          Primary.unfinalise hasFS err lastValidEpoch
+
+-- | Validate the given most recent epoch. If that epoch contains no valid
+-- block, try the epoch before it, and so on. Stop as soon as an epoch with a
+-- valid block is found, returning that epoch and the tip corresponding to
+-- that block. If no valid blocks are found, epoch 0 and 'TipGen' is returned.
+validateMostRecentEpoch
+  :: forall m hash h e. (IOLike m, Eq hash, HasCallStack)
+  => ValidateEnv m hash h e
+  -> EpochNo
+     -- ^ Most recent epoch on disk, the epoch to validate
+  -> m (EpochNo, ImmTipWithHash hash)
+validateMostRecentEpoch validateEnv@ValidateEnv { hasFS } = go
+  where
+    go :: EpochNo -> m (EpochNo, ImmTipWithHash hash)
+    go epoch = runExceptT
+      (validateEpoch validateEnv ShouldNotBeFinalised epoch Nothing) >>= \case
+        Right (Just validBlk) -> do
+            -- Found a valid block, we can stop now.
+            removeFilesStartingFrom hasFS (epoch + 1)
+            return (epoch, Tip validBlk)
+        _  -- Epoch should be truncated or is empty
+          | epoch > 0 -> go (epoch - 1)
+          | otherwise -> do
+            -- Found no valid blocks on disk
+            removeFilesStartingFrom hasFS 0
+            return (0, TipGen)
+
+-- | Iff the epoch is the most recent epoch, it should not be finalised.
+--
+-- With finalising, we mean: if there are one or more empty slots at the end
+-- of the epoch, the primary index should be padded with offsets to indicate
+-- that these slots are empty. See 'Primary.backfill'.
+data ShouldBeFinalised
+  = ShouldBeFinalised EpochSize
+  | ShouldNotBeFinalised
+  deriving (Show)
+
+-- | Validate the given epoch
+--
+-- * Invalid or missing epoch files will cause truncation. All blocks after a
+--   gap in blocks (due to a missing blocks or invalid block(s)) are
+--   truncated.
+--
+-- * Epoch files are the main source of truth. Primary and secondary index
+--   files can be reconstructed from the epoch files using the
+--   'EpochFileParser'. If index files are missing, corrupt, or do not match
+--   the epoch files, they are overwritten.
+--
+-- * The 'EpochFileParser' checks whether the hashes (header hash) line up
```

Comment:
Include in the documentation of `EpochFileParser` that it should check that the hashes line up.